### PR TITLE
Refs [TASK][34912] Handle event Search screen

### PR DESCRIPTION
**Pull request by manhhung23399**
## Related Tickets
- [#Task 34912: Handle event Search screen](https://edu-redmine.sun-asterisk.vn/issues/34912)
## WHAT
- Create request to themoviedb api to get data(top rated) which parse to models then pour data from remote data
source into RecyclerView
- Click item Chip: show fragment Genres
- Get data from SearchBar create request to themoviedb api to get data(movie by name) which parse to models then
pour data from remote data source into RecyclerView
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/example/moviews/data/MovieDataSource.kt [hunk 1]
```diff
@@ -14,5 +14,6 @@ interface MovieDataSource {
         fun getUpcomingMovie(callback: OnLoadDataCallback<MutableList<Movie>>)
         fun getNowPlayingMovie(callback: OnLoadDataCallback<MutableList<Movie>>)
         fun getPopularMovie(callback: OnLoadDataCallback<MutableList<Movie>>)
+        fun getTopSearches(callback: OnLoadDataCallback<MutableList<Movie>>)
```
Comment (Domhoan) [inline app/src/main/java/com/example/moviews/data/MovieDataSource.kt]:
```suggestion
fun searchTopMovie(callback: OnLoadDataCallback<MutableList<Movie>>)
```

File: app/src/main/java/com/example/moviews/data/remote/MovieRemoteDataSource.kt [hunk 2]
```diff
@@ -9,7 +9,7 @@ import com.example.moviews.utils.Constant
 
 class MovieRemoteDataSource : MovieDataSource.Remote {
 
-    private fun builUrl(paths: List<String>): String {
+    private fun buildUrl(paths: List<String>): String {
         val builder = Uri.Builder()
         builder.scheme(Constant.BASE_HTTPS)
             .authority(Constant.BASE_AUTHORITY)
```
Comment (Domhoan) [inline app/src/main/java/com/example/moviews/data/remote/MovieRemoteDataSource.kt]:
```suggestion
private fun buildUrl(paths: List<String>) = Uri.Builder()
        .scheme(Constant.BASE_HTTPS)
        .authority(Constant.BASE_AUTHORITY)
        ....
        .toString()
```

File: app/src/main/java/com/example/moviews/data/remote/SearchRemoteDataSource.kt [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        builder.scheme(Constant.BASE_HTTPS)
+            .authority(Constant.BASE_AUTHORITY)
+            .appendPath(Constant.BASE_VERSION)
+            .apply {
+                paths.forEach {
+                    this.appendPath(it)
+                }
+            }
+            .appendQueryParameter(Constant.BASE_KEY, BuildConfig.API_KEY)
+        return builder.toString()
```
Comment (Domhoan) [inline app/src/main/java/com/example/moviews/data/remote/SearchRemoteDataSource.kt]:
```suggestion
private fun buildUrl(paths: List<String>) = Uri.Builder().scheme(Constant.BASE_HTTPS)
        .authority(Constant.BASE_AUTHORITY)
        .appendPath(Constant.BASE_VERSION)
        .apply {
            paths.forEach {
                this.appendPath(it)
            }
        }
        .appendQueryParameter(Constant.BASE_KEY, BuildConfig.API_KEY)
        .toString()
```

File: app/src/main/java/com/example/moviews/data/model/Genres.kt [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package com.example.moviews.data.model
+
+import org.json.JSONObject
+
+data class Genres(
```
Comment (Domhoan) [inline app/src/main/java/com/example/moviews/data/model/Genres.kt]:
```suggestion
data class Genre(
```

File: app/src/main/java/com/example/moviews/data/remote/SearchRemoteDataSource.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        GetJsonFromUrl(Genres.GENRES, callback).execute(urlGenres)
+    }
+
+    override fun getTopSearches(callback: OnLoadDataCallback<MutableList<Movie>>) {
+        GetJsonFromUrl(Movie.MOVIE_RESULTS, callback).execute(urlTopSearch)
+    }
+
+    override fun searchMovie(name:String,callback: OnLoadDataCallback<MutableList<Movie>>) {
+        GetJsonFromUrl(Movie.MOVIE_RESULTS,callback).execute(urlSearchMovie+Constant.BASE_QUERY+name)
+    }
```
Comment (Domhoan) [inline app/src/main/java/com/example/moviews/data/remote/SearchRemoteDataSource.kt]: fomat code

File: app/src/main/java/com/example/moviews/repository/RepositoryUtils.kt [hunk 6]
```diff
@@ -15,4 +17,8 @@ object RepositoryUtils {
         val remote = MovieRemoteDataSource.getInstance()
         return MovieRepository.getInstance(remote,local)
     }
+    fun getSearchRepository():SearchRepository{
+        val remote = SearchRemoteDataSource.getInstance()
+        return SearchRepository.getInstance(remote)
+    }
```
Comment (thonc-2135) [inline app/src/main/java/com/example/moviews/repository/RepositoryUtils.kt]: chưa format nè , có thể viết gọn lại thì viết

File: app/src/main/java/com/example/moviews/screen/search/adapter/TopSearchAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        val view = LayoutInflater.from(parent.context)
+            .inflate(R.layout.item_movie_search, parent, false)
+        return TopSearchViewHolder(view, onClickItem)
+    }
+
+    override fun onBindViewHolder(holder: TopSearchViewHolder, position: Int) {
+        holder.bindViewData(topSearches[position])
+    }
+
+    override fun getItemCount(): Int =topSearches.size
```
Comment (thonc-2135) [inline app/src/main/java/com/example/moviews/screen/search/adapter/TopSearchAdapter.kt]: chưa format đâu đấy
```suggestion
    override fun getItemCount() = topSearches.size
```

File: app/src/main/java/com/example/moviews/screen/search/SearchPresenter.kt [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    override fun searchMovie(name:String){
+        repository.searchMovie(name,object :OnLoadDataCallback<MutableList<Movie>>{
+            override fun onSuccess(data: MutableList<Movie>) {
+                view.showFoundedMovie(data)
+            }
+
+            override fun onError(e: Exception) {
+                view.showError(e)
+            }
+
```
Comment (thonc-2135) [inline app/src/main/java/com/example/moviews/screen/search/SearchPresenter.kt]: remove blank

File: app/src/main/java/com/example/moviews/screen/search/SearchPresenter.kt [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+            override fun onError(e: Exception) {
+                view.showError(e)
+            }
+        })
+    }
+    override fun searchMovie(name:String){
+        repository.searchMovie(name,object :OnLoadDataCallback<MutableList<Movie>>{
+            override fun onSuccess(data: MutableList<Movie>) {
+                view.showFoundedMovie(data)
```
Comment (thonc-2135) [inline app/src/main/java/com/example/moviews/screen/search/SearchPresenter.kt:42]: showFoundedMovie là cái gì thế

File: app/src/main/java/com/example/moviews/screen/search/SearchPresenter.kt [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    override fun getGenres() {
+        repository.getGenres(object : OnLoadDataCallback<MutableList<Genre>> {
+            override fun onSuccess(data: MutableList<Genre>) {
+                view.showGenres(data)
+            }
+
+            override fun onError(e: Exception) {
+                view.showError(e)
+            }
+
```
Comment (thonc-2135) [inline app/src/main/java/com/example/moviews/screen/search/SearchPresenter.kt]: remove blank

File: app/src/main/java/com/example/moviews/screen/search/SearchPresenter.kt (hunk 9 above)
Comment (manhhung23399) [reply app/src/main/java/com/example/moviews/screen/search/SearchPresenter.kt:42]: là show phim tìm được ạ

**Pull request merged**